openwisp_controller/connection/connectors/ssh.py

```diff
@@ -0,0 +1,132 @@
+import logging
+import socket
+import sys
+from io import BytesIO
+
+import paramiko
+from django.utils.functional import cached_property
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError as SchemaError
+from scp import SCPClient
+
+if sys.version_info.major > 2:  # pragma: nocover
+    from io import StringIO
+else:  # pragma: nocover
+    from StringIO import StringIO
+
+
+logger = logging.getLogger(__name__)
+SSH_CONNECTION_TIMEOUT = 5
+SSH_AUTH_TIMEOUT = 2
+SSH_COMMAND_TIMEOUT = 5
+
+
+class Ssh(object):
+    schema = {
```

Comment:
@atb00ker this is the schema for SSH credentials. In the future we may have other ways to access devices. For example, `OpenWrt` has an HTTP-RPC API served by `ubus`. Ubiquiti AirOS also has some sort o HTTP access that allows to update the config via HTTP calls (not sure if they have an API but surely calls to their web interface can be simulated).
That's where we want to get at some point with https://github.com/openwisp/netjsonconfig/pull/91

Reply:
That seems great. I see a lot of things i need to learn about and i'll start with reading the same.
Thanks. :smile: